CONTRIBUTING.md

```diff
@@ -0,0 +1,423 @@
+- The `fabric-lifecycle-events-v1` module is a good example of event naming standards.
+
+
+#### Event ordering
+
+- Processes happening in multiple steps should use multiple events.
+    - While Fabric provides an event phase system, purpose-driven events should always be preferred.
+    - For example, an event that both cancels and notifies of an action will produce false notifications (notified but later canceled). It would be preferable to have an `AllowXxx` event for cancellation, and then a `BeforeXxx` event for notification.
+
+## Implementation guidelines
+
+### Simplicity
+
+- Simple code that is easy to debug and reason about is generally preferable to the shortest possible implementation.
+- Limited duplication can be better than indirection, unless the code is complex or used several times.
+- Indirections might make the code harder to read, and should be weighed against their benefits. Examples include:
+    - Lambda composition (forEach, streams).
+    - Method calls.
+    - Method splitting.
+    - Recursion.
+    - Complex class hierarchies.
+- Functional operators (i.e. interfaces) should be designed for maximal readability, and only used when necessary.
+    - Ask for concrete objects if applicable.
+    - Don't include superfluous methods, for example replace shouldApply+apply with a single apply that returns success.
+    - Consider having the user implement a larger interface as a class.
+    - Consider using a custom interface if it is beneficial for comprehension or documentation.
+- The overall complexity of a module shouldn't be ignored in favor of the simplicity of each individual piece. Similarly, the overall complexity of one small piece shouldn't be ignored in favor of the simplicity of the module as a whole.
+
+### Code quality
+- Strongly validate inputs to detect misuse.
+    - Exceptions should be thrown in the first method that can reasonably detect misuse.
+    - Precondition assertions such as `Objects.requireNonNull` for non-null parameters are strongly encouraged.
+    - This might apply to other cases, such as strong JSON validation.
+- Use the weakest suitable interface or class when exposing anything without harming expected uses.
+    - However, prefer `Collection` over `Iterable`.
+- Non-trivial reused (directly or derived) inline constants should be moved to static final fields.
+- Javac shouldn't produce any warnings in its default configuration.
+
+### Thread safety
+
+Since Minecraft has two primary threads, the render and server threads, APIs dealing with shared state need to consider thread safety. This will usually require consideration on a case-by-case basis.
+
+- A cache shared between the render and the server threads may for example wish to use locking for infrequent accesses, or otherwise a copy-on-write strategy to ensure lock-free reads and thread-safe writes.
+- Read-only access from multiple threads does not need synchronization.
```

Comment:
I hope you add another entry for lazy initialization (like benign race), such as:
```java
private Stuff cachedStuff;

private Stuff stuff() {
  var stuff = this.cachedStuff;
  if (stuff != null) return stuff;
  return this.cachedStuff = compute();
}
```

Frequently, people will write code like
```java
private Stuff stuff() {
  if (this.cachedStuff != null) return this.cachedStuff;
  return this.cachedStuff = compute();
}
```
which reads `cachedStuff` field twice and is subject to races, as the first read may be non-null but the second one may.

Reply:
Both are correct provided that `cachedStuff` is `volatile`. The reason we use an intermediate variable is to avoid two `volatile` reads, which are more expensive than a single `volatile` read. I'll add something in that area.

Reply:
Are there actually APIs where this matters, btw?

Reply:
This is a point of concern in the API implementation, especially that there are server and client threads

Reply:
I don't see APIs with this use pattern intended to be called on both sides, but guess there can be.

Reply:
You can find it in the depths of the transfer API: https://github.com/FabricMC/fabric/blob/bdca8858a1b0d5e7885081a27d92e8c3754a9c03/fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/item/ItemVariantImpl.java#L144-L153

It's also used in the content registries:
https://github.com/FabricMC/fabric/blob/bdca8858a1b0d5e7885081a27d92e8c3754a9c03/fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/FlammableBlockRegistryImpl.java#L49-L70
https://github.com/FabricMC/fabric/blob/bdca8858a1b0d5e7885081a27d92e8c3754a9c03/fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/FuelRegistryImpl.java#L51-L59